include/smack/smack.h

```diff
@@ -135,32 +224,143 @@ void __SMACK_decls() {
   D("axiom (forall i: int :: $fp2si($si2fp(i)) == i);");
   D("axiom (forall f: float :: $si2fp($fp2si(f)) == f);");
 
-  // Memory Model
-  D("function $base(int) returns (int);");
+#ifdef BITVECTOR
+  D("type i8 = bv8;");
+  D("type i16 = bv16;");
+  D("type i32 = bv32;");
+  D("type i64 = bv32;");
+  D("type ref = i64;");
+  D("type size_t = i32;");
+#else
+  D("type i8 = int;");
+  D("type i16 = int;");
+  D("type i32 = int;");
+  D("type i64 = int;");
+  D("type ref = int;");
+  D("type size_t = int;");
+#endif
 
-  D("const unique $NULL: int;");
-  D("axiom $NULL == 0;");
+  // Memory Model
   D("const $UNDEF: int;");
+  D("function $base(ref) returns (ref);");
+  D("const unique $NULL: ref;");
+#ifdef BITVECTOR
+  D("axiom $NULL == 0bv32;");
+  D("function {:inline} $pa(pointer: ref, index: ref, size: ref) returns (ref) {$add.i32(pointer, $mul.i32(index, size))}");
```

Comment:
The interesting part of getelementptr is that index and size are not necessarily of same type. For most of programs in regression test, index is i64 while size is i32. It's driving me crazy!